src/allmydata/client.py

```diff
@@ -582,6 +639,136 @@ def init_storage(self):
+        """
+        Cause each storage plugin to instantiate its storage server and return
+        them all.
+
+        :return: A ``Deferred`` that fires with storage servers instantiated
+            by all of the given storage server plugins.
+        """
+        return defer.gatherResults(
+            list(
+                plugin.get_storage_server(
+                    self._get_storage_plugin_configuration(plugin.name),
+                    lambda: self.getServiceNamed(StorageServer.name),
+                ).addCallback(
+                    partial(
+                        self._add_to_announcement,
+                        {u"name": plugin.name},
+                    ),
+                )
+                for plugin
+                # The order is fairly arbitrary and it is not meant to convey
+                # anything but providing *some* stable ordering makes the data
+                # a little easier to deal with (mainly in tests and when
+                # manually inspecting it).
+                in sorted(plugins, key=lambda p: p.name)
+            ),
+        )
+
+
+    def _add_to_announcement(self, information, announceable_storage_server):
+        """
+        Create a new ``AnnounceableStorageServer`` based on
+        ``announceable_storage_server`` with ``information`` added to its
+        ``announcement``.
+        """
+        updated_announcement = announceable_storage_server.announcement.copy()
+        updated_announcement.update(information)
+        return AnnounceableStorageServer(
+            updated_announcement,
+            announceable_storage_server.storage_server,
+        )
+
+
+    def _get_storage_plugin_configuration(self, storage_plugin_name):
+        """
+        Load the configuration for a storage server plugin with the given name.
+
+        :return dict: The matching configuration.
+        """
+        try:
+            # Need to reach past the Tahoe-LAFS-supplied wrapper around the
+            # underlying ConfigParser...
+            config = self.config.config.items(
+                "storageserver.plugins." + storage_plugin_name,
+            )
+        except NoSectionError:
+            config = []
+        return dict(config)
+
+
+    def _enable_storage_servers(self, announceable_storage_servers):
+        """
+        Register and announce the given storage servers.
+        """
+        for announceable in announceable_storage_servers:
+            self._enable_storage_server(announceable)
+
+
+    def _enable_storage_server(self, announceable_storage_server):
+        """
+        Register and announce a storage server.
+        """
+        furl_file = self.config.get_private_path(
+            "storage-plugin.{}.furl".format(
+                # Oops, why don't I have a better handle on this value?
+                announceable_storage_server.announcement[u"name"],
+            ),
+        )
+        furl = self.tub.registerReference(
```

Comment:
Hrmm, it's too bad this ends up using a filename (instead of, e.g., `_Config.write_private_config` or so) .. but I guess this is because Foolscap needs the furl in a file..?

Reply:
I agree it's too bad.  I guess we *could* avoid it without changing Foolscap... we just have to do the furl persistence ourselves instead of letting Foolscap do it.  That's probably better, actually, I'll try it.

Reply:
Yea.  That worked.